web/src/LogStore.ts

```diff
@@ -0,0 +1,191 @@
+// Client-side log store, which helps client side rendering and filtering of logs.
+//
+// Looslely adapted from the data structures in
+// pkg/model/logstore/logstore.go
+// but with better support for incremental updates and rendering.
+
+type LogSpan = {
+  manifestName: string
+  lastSegmentIndex: number
+  firstSegmentIndex: number
+}
+
+class LogSegment {
+  spanId: string
+  time: string
+  text: string
+  continuesLine: boolean
+
+  constructor(seg: Proto.webviewLogSegment) {
+    this.spanId = seg.spanId || ""
+    this.time = seg.time || ""
+    this.text = seg.text || ""
+    this.continuesLine = false
+  }
+
+  startsLine() {
+    return !this.continuesLine
+  }
+
+  isComplete() {
+    return this.text[this.text.length - 1] == "\n"
+  }
+}
+
+class LogStore {
+  spans: { [key: string]: LogSpan }
+  segments: LogSegment[]
+
+  constructor() {
+    this.spans = {}
+    this.segments = []
+  }
+
+  append(logList: Proto.webviewLogList) {
+    let newSpans = logList.spans as { [key: string]: Proto.webviewLogSpan }
```

Comment:
Why do we cast this instead of handling the case where `logList.spans` isn't defined?

Reply:
logList.spans is always defined. this isn't about runtime behavior. this is about compile-time behavior. The type definition is "object" and needs to be cast to the correct type. i don't know why the typescript definitions use "object"